app/controllers/multifactor_auths_controller.rb

```diff
@@ -20,17 +20,22 @@ def create
       redirect_to edit_settings_url
     else
       flash[:success] = t(".success")
+      @continue_path = session.fetch("mfa_redirect_uri", edit_settings_path)
+      session.delete("mfa_redirect_uri")
       render :recovery
     end
   end
 
   def update
+    continue_path = edit_settings_path
     if current_user.otp_verified?(otp_param)
       handle_new_level_param
+      continue_path = session.fetch("mfa_redirect_uri", continue_path)
+      session.delete("mfa_redirect_uri")
     else
       flash[:error] = t("multifactor_auths.incorrect_otp")
     end
-    redirect_to edit_settings_url
+    redirect_to continue_path
```

Comment:
nit: This is a tad confusing to me, mostly because of the term `continue_path` throws me off. For the error case it'll always redirect to `edit_settings_url`, but if it's successful, it would be `session[:mfa_redirect_uri]` or fallback to `edit_settings_url`. To make it more clear, I would prefer to have the redirection in each clause rather than holding a temp var continue path.

```
if current_user.otp_verified?(otp_param)
  ...
  redirect to session[:mfa_redirect_uri] or fallback to edit_settings_url
else
  render error
  redirect to edit_settings_url
```